fineract-core/src/main/java/org/apache/fineract/organisation/monetary/domain/ApplicationCurrencyRepositoryWrapper.java

```
@@ -66,4 +63,17 @@ public ApplicationCurrency findOneWithNotFoundDetection(final String currencyCod
         }
         return applicationCurrency;
     }
+
+    public Boolean existsByCode(String code) {
+        return repository.existsByCode(code);
+    }
+
+    public ApplicationCurrency save(ApplicationCurrency currency) {
+        if (existsByCode(currency.getCode())) {
```

Please do the validation as one of the first step in the service! Validation errors should be raised ASAP!

---

Also, code is unique, so you dont event need to check. After it got saved, you would got a `duplicate key violates unique constraint`, you can handle that and raise the proper exception.

---

Yes, but I want to avoid the sql raising an exception, the application should instead raise an exception. I will however move this code to service.

---

you can do it, but pretty unnecessary... however this API will not be under heavy load ever, so it does not have performance impact. also sql exceptions can be wrapped into any "new" exception also... but like i said... it could be fine as this will be used rarely